Кравченко Юлии ПИ19-4
Вариант 12 

Задание 6. Задача об оптимальном использовании ресурсов

Для производства трех видов изделий (A, B, C) используются ресурсы типа I и II,
причем закупки ресурсов ограничены возможностями поставщиков. Нормы расхода ресурсов и их запасы приведены в таблице.


Условие для подстановки:

x1+3*x2+4*x3<=3000

6*x1+5*x2+2*x3<=3320

x1,x2,x3>=0

Целевая функция:

(6*b+12)*x1+(5*b+22)*x2+c*x3

In [7]:
#Модель задачи
import pandas as pd
import scipy.optimize
import numpy as np
from scipy.optimize import minimize
from sympy import *
import math 
import sympy as sym
from scipy import optimize,arange
import matplotlib.pyplot as plt
from sympy import symbols, Function
import sympy as sm
import pandas as pd
x1,x2,x3,p1,p2,b,c,a,a1,b1,c1 = symbols('x1 x2 x3 p1 p2 b c a a1 b1 c1 ') 
F = eval(input("Введите функцию спроса F, например  (6*b+12)*x1+(5*b+22)*x2+c*x3:"))
a1 = int(input("a"))
b1 = int(input("b"))
c1 = int(input("c"))
p1 = int(input("p1")) # запасы 3000
p2 = int(input("p2")) # запасы 3320
F = F.subs([(a,a1),(b,b1),(c,c1)])
print("Функция для максимизации:",F)


Функция для максимизации: 24*x1 + 32*x2 + 19*x3


Построение модели задачи:

In [11]:
data1 = [{'x1':24,'x2':32, 'x3':19},{'x1':1,'x2':3, 'x3':a1, 'Запасы':"<=30000"},{'x1':6,'x2':5, 'x3':2,'Запасы':"<=3320"}]
df = pd.DataFrame(data1, index=['Maximize', 'A','B'])
print(df)

          x1  x2  x3   Запасы
Maximize  24  32  19      NaN
A          1   3   2  <=30000
B          6   5   2   <=3320


In [14]:
import scipy
from scipy.optimize import linprog 

c = [-24,-32,-19] # список коэффициентов функции цели
b_ub = [3000,3320] 
A_ub = [[1,3,a1],
              [6,5,2]]
d=linprog(c, A_ub, b_ub,method='simplex' ) 
#print(d)
for key,val in d.items():
         print(key,val) 
         if key=='x':
                  q=[sum(i) for i in A_ub*val]
                  print('A_ub*x',q) #использованные ресурсы
                  q1= scipy.array(b_ub)-scipy.array(q) 
                  print('b_ub-A_ub*x', q1) #остатки ресурсов


x [  64.    0. 1468.]
A_ub*x [3000.0, 3320.0]
b_ub-A_ub*x [0. 0.]
fun -29428.0
slack [0. 0.]
con []
status 0
message Optimization terminated successfully.
nit 4
success True


C:\Users\kjd20\AppData\Local\Temp/ipykernel_8088/3149713682.py:16: DeprecationWarning: scipy.array is deprecated and will be removed in SciPy 2.0.0, use numpy.array instead
  q1= scipy.array(b_ub)-scipy.array(q)


Найден оптимальный план по видам продукции: x [  64.    0. 1468.]

Найдено фактическое иcпользование ресурсов: A_ub*x [3000.0, 3320.0]

Остатки ресурсов при обеспечении максимальной прибыли:  slack [0. 0.]

Дефицит (исчерпанность ресурсов)= b_ub-A_ub*x [0. 0.]

Оптимальное значение целевой функции 29428.0

С помощью двойственной задачи найду ценность данных ресурсов в отношении прибыли от реализации выпускаемой продукции. Двойственная задача является задачей  минимизации, а Коэффициенты функции цели в прямой задаче являются ограничениями в двойственной задаче.

Двойственная задача

In [13]:
import scipy
from scipy.optimize import linprog
A_ub =[[1,3,a1],
              [6,5,2]] 
c=[3000,3320]
b_ub = [-24,-32,-19]
A_ub_T =-scipy.transpose(A_ub)
d=linprog(c, A_ub_T, b_ub, method='simplex')
print(d)


C:\Users\kjd20\AppData\Local\Temp/ipykernel_8088/3298817082.py:7: DeprecationWarning: scipy.transpose is deprecated and will be removed in SciPy 2.0.0, use numpy.transpose instead
  A_ub_T =-scipy.transpose(A_ub)


     con: array([], dtype=float64)
     fun: 29428.0
 message: 'Optimization terminated successfully.'
     nit: 3
   slack: array([0. , 2.3, 0. ])
  status: 0
 success: True
       x: array([6.6, 2.9])


Как можно видеть наибольшую ценность имеет 1 вид ресурсов, поэтому его надо закупать в первую очередью